### ISSUE TITLE
`NOTIFICATION_PREDELETE` isn't immediately sent when a tool script node is removed

### ISSUE THREAD

**Issue by ershn** [godotengine/godot#84366]
### Godot version

v4.1.2.stable.mono.official [399c9dc39]

### System information

Godot v4.1.2.stable.mono - Windows 10.0.19045 - Vulkan (Forward+) - dedicated Radeon RX 580 Series (Advanced Micro Devices, Inc.; 31.0.21902.5) - AMD Ryzen 7 2700X Eight-Core Processor (16 Threads)

### Issue description

When you remove a node with a tool script attached from a scene, `NOTIFICATION_PREDELETE` isn't sent immediately. You have to close the scene for that notification to be sent to the node.

I'm not entirely sure if it's a bug or if it's intended but it makes using `NOTIFICATION_PREDELETE` in tool scripts pretty hard since you can't use it to run logic for when the node is removed, whether it's because the node was removed manually or because the scene was closed. You have to resort to other signals instead (e.g. `NOTIFICATION_EXIT_TREE`), which don't have the exact same use as `NOTIFICATION_PREDELETE`.

I haven't look at the implementation but it might also result in removed nodes pilling up in memory until the scene is closed.

## Expected behavior

Removing a tool script node from the scene should immediately trigger `NOTIFICATION_PREDELETE`.

### Steps to reproduce

The MRP contains a scene containing a single tool script. That tool script prints 'predelete' when it receives `NOTIFICATION_PREDELETE`.

1. Open the MRP
2. Remove the `Tool` node from the scene using the contextual menu
    - nothing is printed to the output
3. Close the scene
    - `predelete` is printed to the output

### Minimal reproduction project

[NOTIFICATION_PREDELETE isn't immediately called when a tool script node is removed.zip](https://github.com/godotengine/godot/files/13239329/NOTIFICATION_PREDELETE.isn.t.immediately.called.when.a.tool.script.node.is.removed.zip)

**Comment by raulsntos** [comment]
`NOTIFICATION_PREDELETE` is sent when the Node is freed, but it doesn't seem like the Node is freed when it's deleted from the scene it's just removed from the scene tree. I think this may be intentional so that you can undo the delete action and get the node back.

Why is using `NOTIFICATION_EXIT_TREE` not an option? If you want to react to the Node being removed from the scene that sounds like the right notification to use. Maybe you could describe your use case. What are you trying to do and why do you need `NOTIFICATION_PREDELETE` to do it?

**Comment by ershn** [comment]
I see ! I didn't think of the undo history. I did some testing in the editor and it does seem to be the reason why.

After thinking about the problem again, I think I was making wrong assumptions about the meaning of `NOTIFICATION_PREDELETE`. To me "node removal from scene" was equal to "node destruction", and I was using `NOTIFICATION_PREDELETE` as such. But even though that's what happens most of the time, I understand it's not necessarily the case now, whether at runtime or in the editor.

---

To answer your questions, I can definitely use `NOTIFICATION_EXIT_TREE` and that's what I ended up using but the reason I was trying to avoid it is because it's also fired when you reparent the node, not only when you remove it. This means the same node can receive the notification several times during its lifetime. Here, I wanted to use `NOTIFICATION_PREDELETE` for cleanup logic that would be triggered on node removal only.

My use case is as follows.
In a child scene that I instantiate in a parent scene, I have a tool script that paints itself to a tilemap found in the parent scene. The tool script uses its position to determine what cell to use in the tilemap. When the child scene (and thus the tool script node) is moved or reparented in the editor, I need to update which cell is painted on. When the child scene is removed I need to clear the cell (note that I can't use scene tiles here because they don't allow me to modify the instantiated scenes in the editor).

I use `NOTIFICATION_TRANSFORM_CHANGED` to detect position changes and reparents. I was trying to use `NOTIFICATION_PREDELETE` to detect node removal only, but since I can't use it for that, I'm using `NOTIFICATION_EXIT_TREE` now instead. Since `NOTIFICATION_TRANSFORM_CHANGED` and `NOTIFICATION_EXIT_TREE` are both fired on reparents the cell erasing logic will always be triggered 2 times on reparents (which requires adapting the code accordingly).

But as I wrote above, I was making wrong assumptions on `NOTIFICATION_PREDELETE` so I'll try to see if there is another way to improve my implementation without using it.
